Source/Core/VideoCommon/VertexLoaderX64.cpp

```diff
@@ -0,0 +1,453 @@
+	}
+
+	const u64 col[2] = {m_VtxDesc.Color0, m_VtxDesc.Color1};
+	for (int i = 0; i < 2; i++)
+	{
+		if (col[i])
+		{
+			data = GetVertexAddr(ARRAY_COLOR + i, col[i]);
+			ReadColor(data, col[i], m_VtxAttr.color[i].Comp, m_VtxAttr.color[i].Elements);
+			m_native_components |= VB_HAS_COL0 << i;
+			m_native_vtx_decl.colors[i].components = 4;
+			m_native_vtx_decl.colors[i].enable = true;
+			m_native_vtx_decl.colors[i].offset = m_dst_ofs;
+			m_native_vtx_decl.colors[i].type = VAR_UNSIGNED_BYTE;
+			m_native_vtx_decl.colors[i].integer = false;
+			m_dst_ofs += 4;
+		}
+	}
+
+	const u64 tc[8] = {
+		m_VtxDesc.Tex0Coord, m_VtxDesc.Tex1Coord, m_VtxDesc.Tex2Coord, m_VtxDesc.Tex3Coord,
+		m_VtxDesc.Tex4Coord, m_VtxDesc.Tex5Coord, m_VtxDesc.Tex6Coord, m_VtxDesc.Tex7Coord,
+	};
+	for (int i = 0; i < 8; i++)
+	{
+		int elements = m_VtxAttr.texCoord[i].Elements + 1;
+		if (tc[i])
+		{
+			data = GetVertexAddr(ARRAY_TEXCOORD0 + i, tc[i]);
+			u8 scaling_exponent = m_VtxAttr.texCoord[i].Frac;
+			ReadVertex(data, tc[i], m_VtxAttr.texCoord[i].Format, elements, elements, scaling_exponent, &m_native_vtx_decl.texcoords[i]);
+			m_native_components |= VB_HAS_UV0 << i;
+		}
+		if (tm[i])
+		{
+			m_native_components |= (VB_HAS_TEXMTXIDX0 | VB_HAS_UV0) << i;
+			m_native_vtx_decl.texcoords[i].enable = true;
+			MOVZX(64, 8, scratch1, MDisp(src_reg, texmatidx_ofs[i]));
+			PXOR(XMM0, R(XMM0));
+			CVTSI2SS(XMM0, R(scratch1));
+			if (tc[i])
+			{
+				m_native_vtx_decl.texcoords[i].components = 3;
+				if (elements == 1)
+					m_dst_ofs += sizeof(float);
+				MOVSS(MDisp(dst_reg, m_dst_ofs), XMM0);
+				m_dst_ofs += sizeof(float);
+			}
+			else
+			{
+				m_native_vtx_decl.texcoords[i].components = 4;
+				m_native_vtx_decl.texcoords[i].offset = m_dst_ofs;
+				m_native_vtx_decl.texcoords[i].type = VAR_FLOAT;
+				m_native_vtx_decl.texcoords[i].integer = false;
+				SHUFPS(XMM0, R(XMM0), 0x45);
+				MOVUPS(MDisp(dst_reg, m_dst_ofs), XMM0);
+				m_dst_ofs += sizeof(float) * 4;
+			}
+		}
+	}
+
+	if (m_VtxDesc.PosMatIdx)
+	{
+		MOVZX(32, 8, scratch1, MDisp(src_reg, posmatidx_ofs));
+		AND(32, R(scratch1), Imm8(0x3F));
+		MOV(32, MDisp(dst_reg, m_dst_ofs), R(scratch1));
+		m_native_components |= VB_HAS_POSMTXIDX;
+		m_native_vtx_decl.posmtx.components = 4;
+		m_native_vtx_decl.posmtx.enable = true;
+		m_native_vtx_decl.posmtx.offset = m_dst_ofs;
+		m_native_vtx_decl.posmtx.type = VAR_UNSIGNED_BYTE;
+		m_native_vtx_decl.posmtx.integer = true;
+		m_dst_ofs += sizeof(u32);
+	}
+
+	// Prepare for the next vertex.
+	ADD(64, R(dst_reg), Imm32(m_dst_ofs));
+	const u8* cont = GetCodePtr();
+	ADD(64, R(src_reg), Imm32(m_src_ofs));
+
+	SUB(32, R(count_reg), Imm8(1));
+	J_CC(CC_NZ, loop_start);
+
+	// Get the original count.
+	POP(32, R(ABI_RETURN));
+
+	if (m_VtxDesc.Position & MASK_INDEXED)
+	{
+		SUB(32, R(ABI_RETURN), R(skipped_reg));
+		RET();
+
+		SetJumpTarget(m_skip_vertex);
+		ADD(32, R(skipped_reg), Imm8(m_VtxDesc.Position == INDEX8 ? 1 : 2));
```

Comment:
This shall always increase by one. There is no need to skip two vertices if the index is 16 bit wide.